### ISSUE TITLE
Mojo::IOLoop::Stream acts weird when writing to a regular file

### ISSUE THREAD

**Issue by declanmalone** [mojolicious/mojo#1401]
* Mojolicious version: 8.23,
* Perl version: v5.20
* Operating system: Linux (Devuan)

### Steps to reproduce the behavior

Constructor takes a "handle" variable. I try three ways to create a suitable one:

    my ($TMP,$tmp);
    # using open
    open $TMP, ">", "/tmp/temp_output" or die "Couldn't create temp file\n";
    # using IO::File
    $TMP = IO::File->new("/tmp/temp_output", O_CREAT|O_WRONLY|O_NONBLOCK);
    # using a file descriptor 
    my $fd = fileno($TMP);
    # Now create 
    $tmp = Mojo::IOLoop::Stream->new($TMP);  # with "handle"
    $tmp = Mojo::IOLoop::Stream->new($fd);      # with file descriptor

    $tmp->on(write => sub {
	my ($s,$bytes) = @_;
	say "sent $bytes";
	     });
    $tmp->start;

    say $TMP "Writing to temp file";

    Mojo::IOLoop->one_tick;
    Mojo::IOLoop->one_tick;
    Mojo::IOLoop->one_tick;

### Expected behavior

No weirdness

### Actual behavior

1,2:  fh from open() or IO::File both give the same message:

    Mojo::Reactor::EV: I/O watcher failed: Mojo::IOLoop::Stream: Bad file descriptor

3: file descriptor gives the message

    Handle is closed

I don't see this behaviour when reading from a regular file, regardless of whether I pass in a file handle or a file descriptor.

**Comment by declanmalone** [comment]
Sorry.. replace "say" above with:

    $tmp->write("Writing to temp file");

Same messages come up.

**Comment by kraih** [comment]
Don't use `Mojo::IOLoop::Stream` for files, it makes no sense.

**Comment by Grinnz** [comment]
It does not do asynchronous I/O so the only use case where a file is like a stream would be watching a file tail, which you may consider something like https://metacpan.org/pod/IO::Async::FileStream (with IO::Async::Loop::Mojo).

**Comment by declanmalone** [comment]
> It does not do asynchronous I/O

The synopsis says "Mojo::IOLoop::Stream - Non-blocking I/O stream." Digging into the code also reveals checks for things like EAGAIN and whatnot. 

> Don't use Mojo::IOLoop::Stream for files, it makes no sense.

A couple of reasons why it did make sense to me:

- there's nothing in the man page to suggest it _shouldn't_ work for files
- I do intend to use sockets or whatever in future, but it's easier if I test with files first (it means that my upgrade path should work without needing code changes)

If this is not a bug (and I'm wrong to say that it _is_ asynchronous), could someone explain to me how a file handle actually _causes_ the Stream module to behave this way? It just seems to go against the basic Unix philosophy that "everything is a file."

**Comment by Grinnz** [comment]
It asynchronously waits for sockets yes. The actual I/O operation of reads and writes with files will block unless you use something like AIO.

Files and sockets are very different in many ways. This is just due to the underlying operating system implementations. Everything may pretend to be a file but it doesn't mean they will all act the same.

**Comment by declanmalone** [comment]
It seems (thought I haven't tested) that I can't use `IO::AIO` because `Mojo::IOLoop::Stream` doesn't use `aio_*` calls. Maybe it gets translated somewhere deeper, though.

I'm still at a loss to understand why `Mojo::IOLoop::Stream` doesn't work with plain files. I'm pretty sure that I've written low-level code that does async writes to regular files. The process seems to be:

1. open with appropriate non-blocking option
2. use regular writes and check for "would block" type errors
3. use select() or similar in an event loop to make progress

As far as I can tell, this is no different from the scenario here. Or so it would seem.

I did another test with opening a fifo for reading in non-blocking mode, then passed that handle to the Stream constructor. Even though there was no writer at the time, the code worked (instead of blocking on open) and then when another process opened the fifo and started writing, I started getting read events. If that sort of esoteric scenario works with no problem, then why the weirdness with regular files?

@Grinnz, You mentioned use cases, and I'll give you another... the "regular" file isn't actually hosted locally, but is accessed over nfs or sshfs. This (more than `tail -f`) is the kind of case that I'm interested in dealing with.

**Comment by fskale** [comment]
@declanmalone,
a while ago i developed a Mail->SMS Gateway by using Mojo::IOLoop::Stream and then Mojo::UserAgent (using post_p) which then will be converted to a SMS. (Using Minion as job queue)
I wrote a simple demo gist to demonstrate Non-Blocking, delayed writing to /dev/STDOUT. (Works with a Filehandle too).
For Reading changed files, i use AnyEvent::Filesys::Notify.
Perhaps it will lead you to a solution to the problem.
My demo gist:
[https://gist.github.com/fskale/b032a6f34c0700e70498b61c05466696](https://gist.github.com/fskale/b032a6f34c0700e70498b61c05466696)

**Comment by kraih** [comment]
There is no reason to ever use `Mojo::IOLoop::Stream` for regular files because regular files are always readable and writable. You're just wasting time by sending them through the event loop, so we won't go out of our way to support something that has no real world applications. `IO::AIO` does something completely different.

Pipes are a completely different subject, and perfectly fine to use with `Mojo::IOLoop::Stream`.

**Comment by declanmalone** [comment]
@kraih. OK, thanks for pointing out that regular file reads/writes never block. Either I'd never known that or forgotten it along the way.

However, I'm still at a loss as to understand how `Mojo::IOLoop::Stream` manages to fail on regular file writes. I'm not looking for a code update to support this, but I do want to know what is happening. What assumption is being broken (and where) by passing in a regular file handle?

I disagree with what you're saying about there being no reason to want to use `Mojo::IOLoop::Stream` on regular files. The reason why _I_ wanted to use it this way is that I'm taking an old program I had [1] and rewriting its interface to make it work better with an external event loop, and also to go from processing regular files to network streams. I chose to do this in Mojo because:

1.  I'm already using Mojolicious for web interfaces, and `Mojo::IOLoop` seems the natural choice for the event framework (I get a "full stack" solution)
2. I like that I don't have to use a separate event loop like EV or AnyEvent
3. The available primitives are simple, easy to understand and sufficient
4. There _seemed_ to be a ready upgrade path from working with files to network streams (without having to rewrite code later)

It's true that sending sending file data through a `Mojo::IOLoop::Stream` is at best a waste of time (cpu cycles) but I don't care about that. I only care about developer time&mdash;_my_ time. And, yeah, that has been wasted in trying to figure out how something that should have been at worst a no-op was throwing errors. At least that's how I see it.

I'm going to close this, since I can work around the problem, but I would really like an explanation as to what is going on in the internals that makes `Mojo::IOLoop::Stream` fail in this case.

Thanks to everyone that already responded, by the way. I've learned some new things.

[1]: `Crypt::IDA`, which effectively just does a RAID-like transform on data streams. I want to build on this and implement a proper network-based file storage cluster with Mojolicious web interfaces and IPC between the front-end, head node and storage nodes.
